### PR TITLE
tlodderstedt/Authorization request with both authorization_details and resources

### LINKED ISSUE #40: Authorization request with both authorization_details and resources
The example in "2.3. Relationship to "resource" parameter" includes

1. An authorization request with `authorization_details` and without `resource`.
2. A token request with `resource`.

This filters elements in `authorization_details` by the `resource`.

I want to know the expected behavior when an authorization request comes with both `authorization_details` and `resource`s.

Should it be treated as if elements whose `locations` don't include any one of `resource`s specified by the authorization request didn't exist from the beginning? Or, shouldn't `resource`s in the authorization request affect `authorization_details`? Note that the latter still has a meaning because the `resource`s are used as the values of audience (`aud`).

### PR DESCRIPTION
**Pull request by tlodderstedt**
resolves #40

### CODE REVIEW

File: main.md [hunk 1]
```diff
@@ -300,11 +300,28 @@ Alternatively, there could be an authorization data type for OpenID Connect. (#o
 
 ## Relationship to "resource" parameter
 
-The request parameter `resource` as defined in [@I-D.ietf-oauth-resource-indicators] indicates to the AS the resource(s) where the client intends to use the access tokens issued based on a certain grant. This mechanism is a way to audience-restrict access tokens and to allow the AS to create resource server specific access tokens. 
+The request parameter `resource` as defined in [@I-D.ietf-oauth-resource-indicators] indicates to the AS the resource(s) where the client intends to use the access tokens issued based on a certain grant. This mechanism is a way to audience-restrict access tokens and to allow the AS to create resource server specific access tokens. The `authorization_details` parameters also allows to designate the audience of a certain authorization details object in the respective `locations` element.
```
Comment (jricher) [inline main.md:303]:
```suggestion
The request parameter `resource` as defined in [@I-D.ietf-oauth-resource-indicators] indicates to the AS the resource(s) where the client intends to use the access tokens issued based on a certain grant. This mechanism is a way to audience-restrict access tokens and to allow the AS to create resource server specific access tokens. The `authorization_details` parameter also allows the client to designate the audience of a certain authorization details object in the respective `locations` element. 
```

File: main.md [hunk 2]
```diff
@@ -300,11 +300,28 @@ Alternatively, there could be an authorization data type for OpenID Connect. (#o
...
+The request parameter `resource` as defined in [@I-D.ietf-oauth-resource-indicators] indicates to the AS the resource(s) where the client intends to use the access tokens issued based on a certain grant. This mechanism is a way to audience-restrict access tokens and to allow the AS to create resource server specific access tokens. The `authorization_details` parameters also allows to designate the audience of a certain authorization details object in the respective `locations` element. 
 
-If a client uses `authorization_details` with `locations` elements and the `resource` parameter in the same authorization request, the `locations` data take precedence over the data conveyed in the `resource` parameter for that particular authorization details object.
+This specification allows a client to use both parameters can be used together in an authorization request and it defines how the `resource` parameter in the token request can be used to assign authorization details to a certain access token.
```
Comment (jricher) [inline main.md:305]:
```suggestion
This specification allows a client to use both parameters together in an authorization request, and it defines how the `resource` parameter in the token request can be used to assign authorization details to a certain access token.
```

File: main.md [hunk 3]
```diff
@@ -300,11 +300,28 @@ Alternatively, there could be an authorization data type for OpenID Connect. (#o
...
```
Comment (jricher) [inline main.md:306]:
```suggestion

If used together, the `locations` element within objects of the `authorization_details` parameter overrides the value of the `resources` parameter. In the absence of a `locations` element, the value of the `resources` parameter is applied to the object.
```

**Review by jricher — CHANGES_REQUESTED** [review]
Some small clean up and I think a bit of expounding on the relationship.

File: main.md [hunk 4]
```diff
@@ -300,11 +300,28 @@ Alternatively, there could be an authorization data type for OpenID Connect. (#o
...
+
+### Token Request
+
+If  a client uses the `resource` parameter in a token requests, the AS MUST utilize the data provided in the `locations` elements to filter the authorization data objects applicable to the respective resource(s). 
+
+The logic is as follows:
+
+* For every authorization details object without a `locations` element: the authorization server treats it as applicable to all resources, i.e. it assigns this authorization details object 
+to the access token. 
+* For every authorization details object with a `locations` element: the authorization server adds this object to the access token, if at least one of one `locations` values excactly matches the `resource` parameter value. The authorization server MUST compare both values using an exact byte match of the string values.
```
Comment (b---c) [inline main.md:324]:
```suggestion
* For every authorization details object with a `locations` element: the authorization server adds this object to the access token, if at least one of the `locations` values exactly matches the `resource` parameter value. The authorization server MUST compare both values using an exact byte match of the string values.
```

**Pull request merged**